exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/StatisticsMergeBatch.java

```diff
@@ -0,0 +1,532 @@
+  // Map of non-map fields to VV in the incoming schema
+  private Map<MaterializedField, ValueVector> copySrcVecMap = null;
+  private Map<String, Map<String, ValueHolder>> aggregationMap = null;
+  public StatisticsMergeBatch(StatisticsMerge popConfig, RecordBatch incoming,
+                              FragmentContext context) throws OutOfMemoryException {
+    super(popConfig, context, incoming);
+    this.functions = new HashMap<>();
+    this.aggregationMap = new HashMap<>();
+
+    /*for (String key : popConfig.getFunctions()) {
+      aggregationMap.put(key, new HashMap<String, ValueHolder>());
+      if (key.equalsIgnoreCase("statcount") || key.equalsIgnoreCase("nonnullstatcount")) {
+        functions.put(key, "sum");
+      } else if (key.equalsIgnoreCase("hll")) {
+        functions.put(key, "hll_merge");
+      } else if (key.equalsIgnoreCase("sum_width")) {
+        functions.put(key, "avg_width");
+      }
+    }*/
+    for (String key : popConfig.getFunctions()) {
+      if (key.equalsIgnoreCase("sum_width")) {
+        functions.put(key, "avg_width");
+      } else if (key.equalsIgnoreCase("hll")) {
+        functions.put(key, "hll_merge");
+      } else {
+        functions.put(key, key);
+      }
+      aggregationMap.put(functions.get(key), new HashMap<String, ValueHolder>());
+    }
+  }
+
+  private void createKeyColumn(String name, LogicalExpression expr, Map<MaterializedField, ValueVector> parentMap)
+      throws SchemaChangeException {
+    ErrorCollector collector = new ErrorCollectorImpl();
+
+    LogicalExpression mle = ExpressionTreeMaterializer.materialize(expr, incoming, collector,
+        context.getFunctionRegistry());
+
+    MaterializedField outputField = MaterializedField.create(name, mle.getMajorType());
+    ValueVector vector = TypeHelper.getNewVector(outputField, oContext.getAllocator());
+    container.add(vector);
+
+    if (collector.hasErrors()) {
+      throw new SchemaChangeException("Failure while materializing expression. "
+          + collector.toErrorString());
+    }
+    parentMap.put(outputField, vector);
+  }
+
+  private ValueVector addMapVector(String name, MapVector parent, LogicalExpression expr) throws SchemaChangeException {
+    ErrorCollector collector = new ErrorCollectorImpl();
+
+    LogicalExpression mle = ExpressionTreeMaterializer.materialize(expr, incoming, collector,
+        context.getFunctionRegistry());
+
+    Class<? extends ValueVector> vvc =
+        TypeHelper.getValueVectorClass(mle.getMajorType().getMinorType(), mle.getMajorType().getMode());
+    ValueVector vector = parent.addOrGet(name, mle.getMajorType(), vvc);
+
+    if (collector.hasErrors()) {
+      throw new SchemaChangeException("Failure while materializing expression. "
+          + collector.toErrorString());
+    }
+    return vector;
+  }
+
+  /**
+   * Identify the list of fields within a map which are unpivoted as columns in output
+   */
+  private void buildKeyList() {
```

Comment:
Renamed `buildKeyList()` to `buildColumnsList()`. Use the "columns" map and verify the rest against it. Also check that only the expected input functions appear in the incoming batch.